modules/log4j-configurator/src/main/java/org/mule/runtime/module/log4j/internal/ApplicationReconfigurableLoggerContextSelector.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.log4j.internal;
+
+import static org.mule.runtime.deployment.model.internal.artifact.CompositeClassLoaderArtifactFinder.findClassLoader;
+
+import static java.lang.ClassLoader.getSystemClassLoader;
+import static java.util.Arrays.asList;
+
+import org.mule.runtime.core.internal.util.CompositeClassLoader;
+import org.mule.runtime.deployment.model.api.policy.PolicyTemplateDescriptor;
+import org.mule.runtime.module.artifact.activation.internal.classloader.MuleSharedDomainClassLoader;
+import org.mule.runtime.module.artifact.api.classloader.ArtifactClassLoader;
+import org.mule.runtime.module.artifact.api.classloader.RegionClassLoader;
+
+import java.net.URI;
+import java.util.List;
+
+import org.apache.logging.log4j.core.LoggerContext;
+import org.apache.logging.log4j.core.selector.ContextSelector;
+
+/**
+ * A {@link ContextSelector} that will use the context for the application once it is deployed. It may be used in environments
+ * that deploys only one app (W-12356154).
+ *
+ * @since 4.7.0
+ */
+public class ApplicationReconfigurableLoggerContextSelector implements ContextSelector {
+
+  private final MuleLoggerContextFactory loggerContextFactory = new MuleLoggerContextFactory();
+  private static final ClassLoader SYSTEM_CLASSLOADER = getSystemClassLoader();
+  private LoggerContext containerLoggerContext;
+
+  private boolean reconfigured;
+
+  public ApplicationReconfigurableLoggerContextSelector() {
+    // The container logger context is created with no logging separation.
+    // This will guarantee that by default this will work as no separation in logs.
+    this.containerLoggerContext =
+        this.loggerContextFactory.build(SYSTEM_CLASSLOADER, this, false);
+  }
+
+  public LoggerContext getContext(String fqcn, ClassLoader loader, boolean currentContext) {
+    return containerLoggerContext;
+  }
+
+  public LoggerContext getContext(String fqcn, ClassLoader loader, boolean currentContext, URI configLocation) {
+    return getContext(fqcn, loader, currentContext);
+  }
+
+  public List<LoggerContext> getLoggerContexts() {
+    return asList(containerLoggerContext);
+  }
+
+  public void removeContext(LoggerContext context) {}
```

Comment:
does something break if this throws an exception?

Reply:
I've added an exception in case it is invoked.